static/src/javascripts/projects/common/modules/commercial/dfp/messenger.js

```diff
@@ -0,0 +1,149 @@
+define([
+    'Promise',
+    'common/utils/report-error'
+], function (Promise, reportError) {
+    var currentHost = location.protocol + '//' + location.host;
+    var listeners = {};
+    var registeredListeners = 0;
+
+    var error400 = { code: 400, message: 'Bad request' };
+    var error405 = { code: 405, message: 'Service %% not implemented' };
+    var error418 = { code: 418, message: 'I\'m a teapot' };
+    var error500 = { code: 500, message: 'Internal server error\n\n%%' };
+
+    return {
+        register: register,
+        unregister: unregister
+    };
+
+    function register(type, callback, _window) {
+        if( registeredListeners === 0 ) {
+            on(_window || window);
+        }
+
+        listeners[type] || (listeners[type] = []);
```

Comment:
This is an improved version of the expression `a = a || b`, which is kind of idiomatic javascript. You can find it in many OS libs.